src/VisualStudio/Core/Def/Implementation/Experimentation/AnalyzerVsixSuggestedActionCallback.cs

```diff
@@ -0,0 +1,234 @@
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Experimentation
+{
+    [Export(typeof(ISuggestedActionCallback))]
+    internal class AnalyzerVsixSuggestedActionCallback : ForegroundThreadAffinitizedObject, ISuggestedActionCallback
+    {
+        private const string AnalyzerEnabledFlight = @"LiveCA/LiveCAcf";
+        private const string AnalyzerVsixHyperlink = @"https://aka.ms/livecodeanalysis";
+        private static readonly Guid FxCopAnalyzersPackageGuid = Guid.Parse("{4A41D270-A97F-4639-A352-28732FC410E4}");
+
+        private readonly VisualStudioWorkspace _workspace;
+        private readonly IForegroundNotificationService _foregroundNotificationService;
+        private readonly IAsynchronousOperationListener _operationListener;
+        private readonly SVsServiceProvider _serviceProvider;
+
+        private readonly object _scheduleLock = new object();
+        // If we are currently checking to see if the user fits the treatment profile, then we don't want to schedule another check
+        private bool _checksRunning = false;
+        // We don't show the user an infobar more than once per VS session
+        private bool _infoBarShown = false;
+        private FxCopInstallStatus _installStatus = FxCopInstallStatus.Unchecked;
+        // Initialized the first time we're on the UI thread
+        private IExperimentationService _experimentationService;
+
+        [ImportingConstructor]
+        public AnalyzerVsixSuggestedActionCallback(VisualStudioWorkspace workspace,
+            IForegroundNotificationService foregroundNotificationService,
+            [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners,
+            SVsServiceProvider serviceProvider)
+        {
+            _workspace = workspace;
+            _foregroundNotificationService = foregroundNotificationService;
+            _operationListener = new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.ABTest);
+            _serviceProvider = serviceProvider;
+        }
+
+        public void OnSuggestedActionExecuted(SuggestedAction action)
+        {
+            AssertIsForeground();
+
+            // Initialize the experimentation service if it hasn't yet been fetched
+            if (_experimentationService == null)
+            {
+                _experimentationService = _workspace.Services.GetRequiredService<IExperimentationService>();
+            }
+
+            // If we haven't yet checked to see if the VSIX is installed, check now
+            if (_installStatus == FxCopInstallStatus.Unchecked)
+            {
+                var vsShell = _serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
+                ErrorHandler.ThrowOnFailure(vsShell.IsPackageInstalled(FxCopAnalyzersPackageGuid, out int installed));
```

Comment:
do we want to crash VS to do this? otherwise, might be fine to just not participate in the AB test for this user.